{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%pip install scikit-allel\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import allel\n",
    "import tabix\n",
    "import sys\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome_number = sys.argv[1]\n",
    "yri=\"NA18486\"\n",
    "gbr=\"HG00148\"\n",
    "\n",
    "yri_index = 1764\n",
    "gbr_index = 54\n",
    "\n",
    "chr_size = pd.read_csv('/home/jbazanwilliamson/Enhancer-Regions-Dplus/size_chr', sep='\\t', dtype=str, names=['chr','start','end'])\n",
    "data_human = tabix.open('/home/jbazanwilliamson/thousandgenomes/chr1.biallelic.vcf.gz')\n",
    "data_archaic = tabix.open('/courses/ehuerta-sanchez/AltaiNea_Den_combined/AltaiNea_Den_combined.1.vcf.gz')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.path.exists('/courses/ehuerta-sanchez/AltaiNea_Den_combined/AltaiNea_Den_combined.1.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-cb120d699cd0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mnew_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdefaultdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdefaultdict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdata_human\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0;32mif\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"0|0\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m54\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1764\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m\"1|1\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m54\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1764\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         new_data[i[1]].update({\"human reference\":i[3],\"human alt\":i[4],\"yri\":i[1764],\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#Get African and European alleles\n",
    "data_human = data_human.query(\"1\",0,249250621)\n",
    "new_data = defaultdict(defaultdict)\n",
    "\n",
    "for i in data_human:\n",
    "    if(\"0|0\" in i[54]+i[1764] and \"1|1\" in i[54]+i[1764]):\n",
    "        new_data[i[1]].update({\"human reference\":i[3],\"human alt\":i[4],\"yri\":i[1764],\n",
    "        \"gbr\":i[54]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Lesly's function to get chimp allele\n",
    "def get_chimp_allele(string):\n",
    "  index=string.find(\"CAnc\")\n",
    "  if index < 0:\n",
    "    return(\"\")\n",
    "  substring=string[index:]\n",
    "  chimp_allele=substring.split(\";\")[0].split(\"=\")[1]\n",
    "  return(chimp_allele)\n",
    "#\n",
    "data_new_archaic = data_archaic.query(\"1\",0,249250621) \n",
    "\n",
    "\n",
    "#Get neanderthal and chimp\n",
    "for spline in data_new_archaic:\n",
    "    if \".\" in spline[9].split(\":\")[0]:\n",
    "      continue\n",
    "    #is it from biallelic human SNPs\n",
    "    if(spline[1] in list(new_data.keys())):\n",
    "            new_data[spline[1]].update({\"archaic reference\":spline[3],\"archaic alt\":spline[4], \"chimp allele\":get_chimp_allele(spline[7]),\"neanderthal\":spline[9].split(\":\")[0]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>yri</th>\n",
       "      <th>gbr</th>\n",
       "      <th>chimp allele</th>\n",
       "      <th>neanderthal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>748061</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758274</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758770</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772028</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>776112</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215121</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215211</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215909</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2220425</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2220517</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        yri gbr chimp allele neanderthal\n",
       "748061    A   T            T           T\n",
       "758274    C   G            G           G\n",
       "758770    A   C            C           C\n",
       "772028    A   G            G           G\n",
       "776112    C   T            T           T\n",
       "...      ..  ..          ...         ...\n",
       "2215121   T   C            C           C\n",
       "2215211   C   T            C           C\n",
       "2215909   C   A            A           A\n",
       "2220425   T   C            C           C\n",
       "2220517   A   C            A           C\n",
       "\n",
       "[252 rows x 4 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Get alleles for the 4 populations\n",
    "df = (pd.DataFrame.from_dict(new_data)).T\n",
    "df['chimp allele'].replace('', np.nan, inplace=True)\n",
    "df['chimp allele'].replace('-', np.nan, inplace=True)\n",
    "df.dropna(subset=['chimp allele'], inplace=True)\n",
    "#Remove heterozygote from neanderthal allele\n",
    "df['neanderthal'].replace('0/1', np.nan, inplace=True)\n",
    "df['neanderthal'].replace('1/0', np.nan, inplace=True)\n",
    "df.dropna(subset=['neanderthal'], inplace=True)  \n",
    "\n",
    "#Change genotype to allele \n",
    "df.loc[df['yri'] == \"0|0\", 'yri'] = df[\"human reference\"]\n",
    "df.loc[df['yri'] == \"1|1\", 'yri'] = df[\"human alt\"]\n",
    "\n",
    "df.loc[df['gbr'] == \"0|0\", 'gbr'] = df[\"human reference\"]\n",
    "df.loc[df['gbr'] == \"1|1\", 'gbr'] = df[\"human alt\"]\n",
    "\n",
    "df.loc[df['neanderthal'] == \"0/0\", 'neanderthal'] = df[\"archaic reference\"]\n",
    "df.loc[df['neanderthal'] == \"1/1\", 'neanderthal'] = df[\"archaic alt\"]\n",
    "\n",
    "\n",
    "df.drop(columns=['human reference', 'human alt', 'archaic reference', 'archaic alt'], inplace=True)\n",
    "df\n",
    "               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get positions that have 2 differents alleles \n",
    "for i, row in df.iterrows():\n",
    "    if len(row.T[row.T.notnull()].unique())!=2:\n",
    "        df.drop([i], inplace=True)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>yri</th>\n",
       "      <th>gbr</th>\n",
       "      <th>chimp allele</th>\n",
       "      <th>neanderthal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>748061</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758274</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758770</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772028</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>776112</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215121</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215211</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215909</th>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2220425</th>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2220517</th>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        yri gbr chimp allele neanderthal\n",
       "748061    B   B            A           B\n",
       "758274    B   B            A           B\n",
       "758770    B   B            A           B\n",
       "772028    B   B            A           B\n",
       "776112    B   B            A           B\n",
       "...      ..  ..          ...         ...\n",
       "2215121   B   B            A           B\n",
       "2215211   B   B            A           B\n",
       "2215909   B   A            A           A\n",
       "2220425   B   B            A           B\n",
       "2220517   A   B            A           B\n",
       "\n",
       "[243 rows x 4 columns]"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Change alleles to A,B frequency\n",
    "\n",
    "df.loc[df['yri'] == df[\"chimp allele\"], 'yri'] = \"A\"\n",
    "df.loc[df['yri'] != df[\"chimp allele\"], 'yri'] = \"B\"\n",
    "\n",
    "df.loc[df['gbr'] == df[\"chimp allele\"], 'gbr'] = \"A\"\n",
    "df.loc[df['gbr'] != df[\"chimp allele\"], 'gbr'] = \"B\"\n",
    "\n",
    "df.loc[df['neanderthal'] == df[\"chimp allele\"], 'neanderthal'] = \"A\"\n",
    "df.loc[df['neanderthal'] != df[\"chimp allele\"], 'neanderthal'] = \"B\"\n",
    "\n",
    "df[\"chimp allele\"]=\"A\"\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "#intermediate file as csv\n",
    "\n",
    "df.to_csv('/home/jbazanwilliamson/Enhancer-Regions-Dplus/chr_1_intermediate.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
